File: apps/desktop/src/app/theme.ts
import type { AppearanceState, ThemeStyleProperties } from "./types";

export function colorWithAlpha(color: string, alpha: number) {
  const hex = color.trim().replace(/^#/, "");
  if (![3, 6].includes(hex.length) || !/^[\da-f]+$/i.test(hex)) {
    return color;
  }

  const expanded = hex.length === 3 ? hex.split("").map((part) => part + part).join("") : hex;
  const red = Number.parseInt(expanded.slice(0, 2), 16);
  const green = Number.parseInt(expanded.slice(2, 4), 16);
  const blue = Number.parseInt(expanded.slice(4, 6), 16);
  return `rgba(${red}, ${green}, ${blue}, ${alpha})`;
}

export function hexColorForPicker(color: string | null | undefined) {
  const value = color?.trim() ?? "";
  return /^#[\da-f]{6}$/i.test(value) ? value : "#caa05d";
}

export function browserLanguages() {
  return navigator.languages?.length ? navigator.languages : [navigator.language || "en-US"];
}

export function activeThemeFromAppearance(appearance: AppearanceState | null) {
  if (!appearance) {
    return null;
  }

  return appearance.themes.find((theme) => theme.id === appearance.activeThemeId && theme.enabled) ?? appearance.themes.find((theme) => theme.enabled) ?? null;
}

export function themeStyleVariables(appearance: AppearanceState | null): ThemeStyleProperties | undefined {
  const theme = activeThemeFromAppearance(appearance);
  if (!theme) {
    return undefined;
  }

  const accent = appearance?.accentOverride ?? theme.tokens.accent;
  return {
    "--surface": theme.tokens.surface,
    "--panel": theme.tokens.panel,
    "--panel-strong": theme.tokens.panelStrong,
    "--text": theme.tokens.text,
    "--muted": theme.tokens.muted,
    "--faint": theme.tokens.faint,
    "--accent": accent,
    "--danger": theme.tokens.danger,
    "--line": theme.tokens.line,
    "--control": theme.tokens.control,
    "--scrollbar-thumb": theme.tokens.scrollbarThumb,
    "--scrollbar-thumb-hover": colorWithAlpha(accent, 0.46),
    "--accent-soft": colorWithAlpha(accent, 0.16),
    "--accent-muted": colorWithAlpha(accent, 0.22),
    "--accent-border": colorWithAlpha(accent, 0.42),
    "--accent-ring": colorWithAlpha(accent, 0.82),
  };
}

File: apps/desktop/src-tauri/src/appearance_store/manifest/theme.rs
use super::primitives::{
    validate_color_token, validate_dotted_identifier, validate_non_empty, validate_simple_semver,
};
use crate::appearance_store::types::{ThemeManifest, ThemeTokens};

pub(super) fn validate_theme_manifest(theme: &ThemeManifest) -> Result<(), String> {
    validate_non_empty("theme id", &theme.id)?;
    validate_non_empty("theme name", &theme.name)?;
    validate_non_empty("theme version", &theme.version)?;
    validate_dotted_identifier("theme id", &theme.id, false)?;
    validate_simple_semver("theme version", &theme.version)?;
    validate_theme_tokens(&theme.tokens)
}

fn validate_theme_tokens(tokens: &ThemeTokens) -> Result<(), String> {
    validate_color_token("surface", &tokens.surface)?;
    validate_color_token("panel", &tokens.panel)?;
    validate_color_token("panelStrong", &tokens.panel_strong)?;
    validate_color_token("text", &tokens.text)?;
    validate_color_token("muted", &tokens.muted)?;
    validate_color_token("faint", &tokens.faint)?;
    validate_color_token("accent", &tokens.accent)?;
    validate_color_token("danger", &tokens.danger)?;
    validate_color_token("line", &tokens.line)?;
    validate_color_token("control", &tokens.control)?;
    validate_color_token("scrollbarThumb", &tokens.scrollbar_thumb)?;
    validate_color_token("scrollbarThumbHover", &tokens.scrollbar_thumb_hover)
}
